### PR TITLE
kv: avoid splitting txns requiring 1PC on retries

### PR DESCRIPTION
**Pull request by spencerkimball**
Previously, on a retry (e.g. possible replay errors) batches would be
split to move the `EndTransactionRequest` into a second part. If the
`Require1PC` flag is set, we now no longer split the batch, which would
simply result in a failure.

Release note: None

### CODE REVIEW

**Review by andreimatei — CHANGES_REQUESTED** [review]
I'm confused about something - if there's an `EndTransaction(commit=true)` in the batch that gets the "possible replay" error, how come the error is retryable and not of the "ambiguous commit" type? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-:-LJnmGAc1q_9gtqNJfFb:b-5aalhn)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-LJnlUU87O0yYyBum1FZ:-LJnlUU87O0yYyBum1F_:b-dcrvil) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ea40e99ad410dbba164dd9f8a8097e8e5c39/pkg/kv/dist_sender.go#L616)):*
> ```Go
> 	splitET := false
> 	var require1PC bool
> 	if et, ok := ba.Requests[len(ba.Requests)-1].GetInner().(*roachpb.EndTransactionRequest); ok && et.Require1PC {
> ```

long line

**Review by spencerkimball — COMMENTED** [review]
@tschottdorf it doesn't fix any open issues I've seen. I'm fixing this because I saw it happen in testing when replicas are lazily initialized (because the timestamp low water mark is set close to the current time).

@andreimatei a possible replay error is retryable.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-:-LJoPq5K4vcJruK4c_P_:b-bft5vl)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-LJnlUU87O0yYyBum1FZ:-LJoQj8g0QlADYEcecpX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ea40e99ad410dbba164dd9f8a8097e8e5c39/pkg/kv/dist_sender.go#L616)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

long line
</blockquote></details>

Done.

**Review by bdarnell — APPROVED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28522)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 656 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-LJoZvx18M3VJir1Ax0g:-LJoZvx18M3VJir1Ax0h:bpscsg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbce6a8d7607a3adf9d9f5265bb1cab807a4101/pkg/kv/dist_sender.go#L656)):*
> ```Go
> 			if len(parts) != 1 {
> 				panic("EndTransaction not in last chunk of batch")
> 			} else if !require1PC {
> ```

I don't think it should be possible to get here with `require1PC`. We only use this flag for point operations, which would only be split if `splitET` were true. We could just panic if `len(parts) != 1 && require1PC`.

**Comment by andreimatei** [comment]
> a possible replay error is retryable

I understand that that's what the code apparently does, but my question is how come that's true when the batch that is detected as being possibly replayed has a commit in it. Is it not possible we already committed in that case (in which case the error should not be retryable)?

**Review by andreimatei — COMMENTED** [review]
I think I can answer my own question: 
a) If the "replay" is caused by a request re-evaluation (`replica.executeWriteBatch()` retry), then the the comment below suggests that, even though the replica creates a retryable error, the batch will not in fact be retried:
https://github.com/cockroachdb/cockroach/blob/2921eb6d5d0b5da4b6ca353e89ed7f5e80728f07/pkg/storage/replica.go#L2545
If it's a re-evaluation, I believe the retryable error is changed into an `AmbiguousResultError` by `replica.executeWriteBatch`.
b) If the "replay" is caused by a DistSender retry, then the DistSender changes the retryable error to a `AmbiguousResultError`.

So, if the client sees a `TransactionRetryError(RETRY_POSSIBLE_REPLAY)`, I believe this means that a lease transfer happened recently or an overloaded timestamp cache bumped its low-water mark; in any case, there's no possible "replay". Which means that the `RETRY_POSSIBLE_REPLAY` is quite bogus.


Would you mind changing the test to create use another reason, not `RETRY_POSSIBLE_REPLAY` - since this one is very confusing (and we should really find a better structure on the server side to not return it)? I think the reason the test uses should not affect the test or your goal here in any way.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-:-LJpgtlTCXQeyUg19sz9:b-id3udz)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by spencerkimball — COMMENTED** [review]
That comment at line (2545) is saying if it's a replay, then sending back a retryable error won't result in any actual retry because the original caller is not going to be there to receive the retryable error and retry.

This is not an ambiguous result error because the batch in this case most certainly has not committed. You're conflating these two cases. The possible replay error is meant to indicate to a caller that it should retry with conviction, because if the caller is there to receive it, it actually means *there was no replay*.

The receipt of `RETRY_POSSIBLE_REPLAY` by a caller means there's wasn't actually a replay. But it's not bogus. In the case where there actually is a replay, the caller won't be there to receive the error, but it was still a replay.

I've added an additional test case with a different error which triggers a txn coordinator retry, which is also good to verify.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-:-LJtCmxoDFOXjraWzBCb:b7qvsfv)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 656 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28522#-LJoZvx18M3VJir1Ax0g:-LJtCasH9K_d3QCXE-yB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbce6a8d7607a3adf9d9f5265bb1cab807a4101/pkg/kv/dist_sender.go#L656)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think it should be possible to get here with `require1PC`. We only use this flag for point operations, which would only be split if `splitET` were true. We could just panic if `len(parts) != 1 && require1PC`.
</blockquote></details>

Done.

**Comment by spencerkimball** [comment]
@andreimatei I think we should have this conversation offline. We seem to be going in circles, slowly. Is there anything still to fix with this PR?

**Comment by andreimatei** [comment]
Don't wait for my review, I'm out this week; I haven't looked at require1PC
to have an opinion.
I haven't changed my mind that saying anything about possible replays in
that test (as well as in the commit message) only serves to confuse readers
(as opposed to talking generally about retriable errors), but this is minor
enough.

**Comment by spencerkimball** [comment]
You need to ok the review for me to merge.

**Review by andreimatei — APPROVED** [review]

**Comment by bdarnell** [comment]
> Approved. Is that a new requirement of some sort?

It's a change on reviewable - it now defaults to setting your github review status to "blocking" in many cases, and when this is set it needs to be explicitly cleared in order to merge. (Spencer, I'm pretty sure there's a way to dismiss the blocking review if you need to)

**Pull request merged**
